lib/posix/pthread_key.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <posix/pthread.h>
+
+struct k_sem sem;
+
+typedef struct pthread_thread_data {
+	sys_dnode_t node;
+
+	/* Key and thread specific data passed to pthread_setspecific() */
+	pthread_key_obj *key;
+	void *spec_data;
+
+} pthread_thread_data;
+
+typedef struct pthread_key_data {
+	sys_dnode_t node;
+	pthread_thread_data thread_data;
+} pthread_key_data;
+
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *new_key;
+
+	*key = NULL;
+
+	k_sem_init(&sem, 1, 1);
+
+	new_key = k_malloc(sizeof(pthread_key_obj));
+
+	if (new_key == NULL) {
+		return ENOMEM;
+	}
+
+	sys_dlist_init(&(new_key->key_data_l));
+
+	new_key->destructor = destructor;
+	*key = (void *)new_key;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *key_obj = (pthread_key_obj *)key;
+	pthread_key_data *key_data;
+	sys_dnode_t *node_l, *next_node_l;
+
+	k_sem_take(&sem, K_FOREVER);
+
+	/* Delete thread-specific elements associated with the key */
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&(key_obj->key_data_l),
+			node_l, next_node_l) {
+
+		/* Remove the object from the list key_data_l */
+		key_data = (pthread_key_data *)
+			sys_dlist_get(&(key_obj->key_data_l));
+
+		/* Deallocate the object's memory */
+		k_free((void *)key_data);
+	}
+
```

Comment:
call destructor here?

Reply:
will leave this for now, not sure if the thread setting key would be calling it at exit

Reply:
I have added the destructor call in `pthread_exit`